feat: implement open zeppelin access add and set cards  functions

Akshola00 opened the pull request:
fixes #105

manlikeHB commented:
Hey @Akshola00, you were supposed to add the `AccessControl` component and restrict those functions to callers who have been granted the admin role.

Akshola00 commented:
> Hey @Akshola00, you were supposed to add the `AccessControl` component and restrict those functions to callers who have been granted the admin role.

GM @manlikeHB 
Does this mean I will create a storage for all admins and implement the access control based on that 
Or is there another way I should implement this

manlikeHB commented:
You have to use OZ's access control component, define the role(s) and grant it in the contract constructor.

see [this](https://github.com/manlikeHB/afk_monorepo/blob/main/onchain/cairo/afk/src/tokens/quest_dn_404.cairo)

Akshola00 commented:
> You have to use OZ's access control component, define the role(s) and grant it in the contract constructor.
> 
> see [this](https://github.com/manlikeHB/afk_monorepo/blob/main/onchain/cairo/afk/src/tokens/quest_dn_404.cairo)
> 
> see how the `mint` function is restricted to the `MINTER_ROLE` [here](self.accesscontrol.assert_only_role(MINTER_ROLE);)

thanks, will make the changes now

Akshola00 commented:
@manlikeHB 
PTAL
implemented the access control for the two specified functions and modified the test to assert that anyone that call these functions are `AUTHORIZED`

manlikeHB commented on onchain/src/contracts/lyricsflip.cairo:
```diff
@@ -73,8 +105,15 @@ pub mod LyricsFlip {
     }
 
+    const ADMIN_ROLE: felt252 = selector!("ADMIN_ROLE");
+    const AUTHORIZED_ROLE: felt252 = selector!("AUTHORIZED_ROLE");
```
No need for `AUTHORIZED_ROLE` role, just the `ADMIN_ROLE` for now.

manlikeHB commented on onchain/src/contracts/lyricsflip.cairo:
```diff
@@ -211,6 +250,7 @@ pub mod LyricsFlip {
 
 
         fn set_cards_per_round(ref self: ContractState, value: u8) {
+            self.accesscontrol.assert_only_role(AUTHORIZED_ROLE);
```
restrict to the `ADMIN_ROLE`

manlikeHB commented on onchain/src/contracts/lyricsflip.cairo:
```diff
@@ -231,6 +271,7 @@ pub mod LyricsFlip {
 
 
         fn add_card(ref self: ContractState, card: Card) {
+            self.accesscontrol.assert_only_role(AUTHORIZED_ROLE);
```
restrict to `ADMIN_ROLE`

manlikeHB commented on onchain/src/contracts/lyricsflip.cairo:
```diff
@@ -286,6 +327,13 @@ pub mod LyricsFlip {
+
+        fn set_role(
+            ref self: ContractState, recipient: ContractAddress, role: felt252, is_enable: bool
+        ) {
```
add test for this function

manlikeHB commented on onchain/src/contracts/lyricsflip.cairo:
```diff
@@ -353,5 +401,17 @@ pub mod LyricsFlip {
+        fn _set_role(
+            ref self: ContractState, recipient: ContractAddress, role: felt252, is_enable: bool
+        ) {
+            self.accesscontrol.assert_only_role(ADMIN_ROLE);
```
restrict to both admin and owner.

manlikeHB commented:
Hi @Akshola00, few changes also CI is failing.

Akshola00 commented:
all changes implemented

manlikeHB commented on onchain/src/contracts/lyricsflip.cairo:
```diff
@@ -365,5 +415,18 @@ pub mod LyricsFlip {
+        ) {
+            self.accesscontrol.assert_only_role(ADMIN_ROLE);
+            self.ownable.assert_only_owner();
+            assert!(role == ADMIN_ROLE, "role not enable");
```
test case for this

Akshola00 commented:
@manlikeHB 
didnt update the test of the newly merged code, pls test now

manlikeHB commented:
Thank you @Akshola00, LGTM!
